File: leetcode/39_Combination_Sum.py
import unittest
from typing import List

class Solution:
    def _rec(self, candidates: List[int], target: int, pos: int, total: int, combination: List[int], combinations: List[List[int]]):
        if total == target:
            combinations.append(combination.copy())
        if total + candidates[pos] <= target:
            for i in range(pos, len(candidates)):
                combination.append(candidates[i])
                self._rec(candidates, target, i, total + candidates[i], combination, combinations)
                combination.pop()

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        combination, combinations = [], []
        self._rec(candidates, target, 0, 0, combination, combinations)
        return combinations

class TestCombinationSum(unittest.TestCase):
    def test_1(self):
        self.assertEqual(Solution().combinationSum([2,3,6,7], 7), [[2,2,3],[7]])    

    def test_2(self):
        self.assertEqual(Solution().combinationSum([2,3,5], 8), [[2,2,2,2], [2,3,3], [3,5]])    

    def test_3(self):
        self.assertEqual(Solution().combinationSum([2], 1), [])    

    def test_4(self):
        self.assertEqual(Solution().combinationSum([1], 1), [[1]])    

    def test_5(self):
        self.assertEqual(Solution().combinationSum([1], 2), [[1, 1]])    

File: leetcode/39_Combination_Sum.rs
struct Solution {}

impl Solution {
    fn _rec(candidates: &Vec<i32>, target: i32, pos: usize, total: i32, combination: &mut Vec<i32>, combinations: &mut Vec<Vec<i32>>) {
        if total == target {
            combinations.push(combination.clone());
        } else if total + candidates[pos] <= target {
            for i in pos..candidates.len() {
                combination.push(candidates[i]);
                Self::_rec(candidates, target, i, total + candidates[i], combination, combinations);
                combination.pop();
            }
        }
    }

    pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
        let mut candidates = candidates;
        let mut combination = Vec::<i32>::new();
        let mut combinations = Vec::<Vec<i32>>::new();
        candidates.sort();
        Self::_rec(&candidates, target, 0, 0, &mut combination, &mut combinations);
        combinations
    }
}

#[cfg(test)]
mod combination_sum_test {
    use super::*;

    #[test]
    fn test_1() {
        assert_eq!(Solution::combination_sum(vec![2,3,6,7], 7), vec![vec![2,2,3], vec![7]]);
    }

    #[test]
    fn test_2() {
        assert_eq!(Solution::combination_sum(vec![2,3,5], 8), vec![vec![2,2,2,2], vec![2,3,3], vec![3,5]]);
    }

    #[test]
    fn test_3() {
        assert_eq!(Solution::combination_sum(vec![2], 1), Vec::<Vec<i32>>::new());
    }

    #[test]
    fn test_4() {
        assert_eq!(Solution::combination_sum(vec![1], 1), vec![vec![1]]);
    }

    #[test]
    fn test_5() {
        assert_eq!(Solution::combination_sum(vec![1], 2), vec![vec![1, 1]]);
    }
}